### ISSUE TITLE
Limit API calls to secrets backend

### ISSUE THREAD

**Issue by itsdalmo** [concourse/concourse#2924]
## What challenge are you facing?

Last month we had ~62 million requests to AWS secrets manager from Concourse. That is about 24 requests per second, every second in a 30 day period. We have 4x ATC's and 8x workers running roughly 400 containers (check and otherwise) at any point in the day, and ~500 secrets used by Concourse that are stored in secrets manager.

Without doing the math, the number of requests seem a bit excessive and represent a significant cost in AWS Secrets Manager, and added costs in Guard Duty for analysing all the secrets manager requests. For SSM I suspect it's very easy to get hit by the rate limit with this amount of requests.

## What would make this better?

The only option that come to mind is implementing a cache to avoid hitting the APIs directly with each request. Also, https://github.com/concourse/concourse/issues/2386 might already help reduce the number of API calls to the secret backend (but I have not read the implementation).

edit: Just read that global resources does not reduce the number of calls to the secret backend according to Vito https://github.com/concourse/concourse/issues/2386#issuecomment-426786169:
> Hmm I don't think this will actually help that case at all. Credential resolution is done prior to even determining whether to run the check or not, since we have to resolve the credentials to determine the resource config to check in the first place.

## Are you interested in implementing this yourself?

I'm interested, but find it hard to navigate the Concourse codebase and would need someone to point me in the right direction 🤗

**Comment by ralekseenkov** [comment]
@marco-m my PR got merged, but it was for reconnects/retries, not for caching.

Concourse has this for Vault (https://concourse-ci.org/creds.html#vault-cache) and not for AWS. I don’t really like how it works and we ended up not using it. It caches secrets for their entire lease duration (which can be pretty long) and makes it hard to rotate them. So if this is enabled we’d need to restart Concourse every time to pick up changes in secrets.

It shouldn’t be too hard to add caching across all cred managers with a configurable duration, in the same way we did reconnects in my PR

@vito ^^ ?

**Comment by AdeOpe** [comment]
We are having this same issue.
So considering grabbing the secrets as the first job in the pipeline and passing it through.
caching would be awesome

**Comment by marco-m** [comment]
@AdeOpe what I suggest is to workaround by deploying your own HashiCorp Vault. All our SSM throttling problems went away. Happy developers, happy ops, no AWS cost (besides the EC2 and the S3 storage for Vault).

**Comment by ralekseenkov** [comment]
the way how caching is done today is definitely suboptimal (if I remember correctly, it's specific to vault credential manager and people also reported issues with caching on a team level vs. pipeline level).

I'm currently looking into another cred mgr related issue - #3331, and will try to take a look at the caching code as well to see if I can make it more generic

**Comment by ralekseenkov** [comment]
https://github.com/concourse/concourse/pull/3628

**Comment by eedwards-sk** [comment]
@ralekseenkov I'm trying out AWS SSM caching on 5.2.0 right now and I'm seeing issues with missing secrets not being retrieved

e.g. if I have a task that wants a non-existent parameter `((foo))`, and I run it, I get:
```
failed to interpolate task config: Expected to find variables: foo
```

if I then set this parameter, and try the task again, I still get the error

I'm only able to get concourse to recognize the value if I destroy / re-create my web and worker nodes

This is with `60m` cache set.

I'm suspecting somehow the cache is preventing it from going to SSM to check the value again, even though it didn't have a cached value from before?

If that's true, is that intended behavior?

Ideally, if a secret wasn't able to be successfully cached, I expect concourse to go to SSM again to retrieve it. I only want to cache secrets that I've successfully retrieved, not a "not found" status.

**Comment by ralekseenkov** [comment]
@eedwards-sk it was the intention. "secret not found" responses are cached in the current implementation. so it will recognize the new value in 60m in your case:
https://github.com/concourse/concourse/blob/master/atc/creds/cached_secrets.go#L54

you can test this by decreasing the duration (e.g. from 60 min to 5 min). it should also be possible to introduce a flag in cache to disable caching of "secret not found" responses, if that becomes a problem for developing new tasks/pipelines when secrets are cached for a long time

@vito FYI

**Comment by eedwards-sk** [comment]
@ralekseenkov 
thanks for confirming the behavior!

I'm curious as to what the advantage of or use case is for caching a "not found" response?

ideally, for what I believe is the common use case, disabling caching of 'secret not found' responses wouldn't be a flag, it would be the default behavior

**Comment by ralekseenkov** [comment]
@eedwards-sk may be you are right -- it could be a default behavior, and there could be a flag to enable caching of "not found" responses. however the problem seems to go beyond that.

* let's say you are developing a new pipeline and you created a secret but put a wrong value by mistake (it happened to me before). it will cached for the entire duration. 
* or let's say you want to rotate a secret and change its value in the underlying cred manager. same story

to some degree this is all very similar to the case you described. this is part of the reason why default cache duration is small, to avoid some of these issues.

**Comment by eedwards-sk** [comment]
@ralekseenkov I definitely foresaw the same issue, which is why I believe we could use a follow-up enhancement that lets us evict the cache either for the entire credential provider or for a specific credential key

otherwise you're basically forcing a lower cache duration just to support credential rotation which starts to defeat the purpose of caching in the first place (especially with SSM where throttling is a big concern)

to summarize, from my testing so far, these are the opportunities for enhancements:

- do not cache 'not found' by default
- allow evicting a cache (could start coarse and eventually be as granular as team/pipeline)

**Comment by ralekseenkov** [comment]
@eedwards-sk I agree with most of the points. however I don't think it defeats the purpose, because for a lot of people (think non-SSM) even 5m cache duration provides value. but I can definitely see the issue and I'm totally up for the follow up enhancement.

we need to figure out how eviction CLI/API can look like and how it would fit into Concourse RBAC.

I think coarse can be a challenge. when you have a large Concourse instance shared between multiple teams, I don't think you'd want to give all your dev teams permissions to drop the entire secret cache.

also... I have to mention this RFC for team-based cred managers which'd change a few things (but might make it easier for teams to drop their own secret cache): https://github.com/concourse/rfcs/pull/21

**Comment by eedwards-sk** [comment]
@ralekseenkov 

>however I don't think it defeats the purpose

Definitely not saying it doesn't have value, the cache addition has HUGE value and is the only reason I'm even testing out SSM now :)

My point is that I don't believe "lower the cache duration" is a proper solution for when you need the cache refreshed (in an out-of-band manner). I think we're both on the same page here.

>I think coarse can be a challenge. when you have a large Concourse instance shared between multiple teams, I don't think you'd want to give all your dev teams permissions to drop the entire secret cache.

Agreed. That suggestion was under the assumption that it might be easier to blow away the whole cache (or whole team's cache) than write the behavior to evict specific keys. Specific keys would be ideal, but I'm happy with "anything that gets us there" if that makes sense.

Thanks for the discourse!

**Comment by ralekseenkov** [comment]
Yes, we are on the same page :) I'll think about it more

**Comment by itsdalmo** [comment]
A way to clear the cache as @eedwards-sk suggests is a great idea, but perhaps we could move it to a separate feature request? I think the original issue is solved with the 5.2.0 release and can be closed?

>  I agree with most of the points. however I don't think it defeats the purpose, because for a lot of people (think non-SSM) even 5m cache duration provides value. but I can definitely see the issue and I'm totally up for the follow up enhancement.

Absolutely, In our case we had 24 requests per second, so a 5min cache saves us several thousand requests to secrets manager in that period and lots of dollars per month as well 😅 Thank you for fixing @ralekseenkov!

**Comment by vito** [comment]
@itsdalmo Yep! I think this should be closed. Let's open a new issue for later iterations. :+1:

**Comment by eedwards-sk** [comment]
thanks all -- opened #3895 and #3896 as follow ups
